src/java.base/share/classes/com/sun/crypto/provider/ML_KEM.java

```diff
@@ -0,0 +1,1509 @@
+    private final int mlKem_k;
+    private final int mlKem_eta1;
+    private final int mlKem_eta2;
+
+    private final int mlKem_du;
+    private final int mlKem_dv;
+    public final int encapsulationSize;
+
+    public ML_KEM(int size) {
+        switch (size) {
+            case 512 -> {
+                mlKem_k = 2;
+                mlKem_eta1 = 3;
+                mlKem_eta2 = 2;
+                mlKem_du = 10;
+                mlKem_dv = 4;
+            }
+            case 768 -> {
+                mlKem_k = 3;
+                mlKem_eta1 = 2;
+                mlKem_eta2 = 2;
+                mlKem_du = 10;
+                mlKem_dv = 4;
+            }
+            case 1024 -> {
+                mlKem_k = 4;
+                mlKem_eta1 = 2;
+                mlKem_eta2 = 2;
+                mlKem_du = 11;
+                mlKem_dv = 5;
+            }
+            default -> throw new IllegalArgumentException(
+                    "Invalid size for ML_KEM-" + size);
+        }
+        mlKem_size = size;
+        encapsulationSize = (mlKem_k * mlKem_du + mlKem_dv) * 32;
+    }
+
+    /*
+    Classes for the internal K_PKE scheme
+     */
+    public record K_PKE_EncryptionKey(byte[] keyBytes) {}
+
+    public record K_PKE_DecryptionKey(byte[] keyBytes) {}
+
+    public record K_PKE_KeyPair(
+            K_PKE_DecryptionKey privateKey, K_PKE_EncryptionKey publicKey) {
+    }
+
+    public record K_PKE_CipherText(byte[] encryptedBytes) {
+    }
+
+    private boolean isValidCipherText(K_PKE_CipherText cipherText) {
+        return (cipherText.encryptedBytes.length == encapsulationSize);
+    }
+
+    /*
+    Classes for internal KEM scheme
+     */
+    public record ML_KEM_EncapsulationKey(byte[] keyBytes) {
+    }
+
+    public record ML_KEM_DecapsulationKey(byte[] keyBytes) {
+    }
+
+    public record ML_KEM_KeyPair(ML_KEM_EncapsulationKey encapsulationKey,
+                                 ML_KEM_DecapsulationKey decapsulationKey) {
+    }
+
+    public record ML_KEM_EncapsulateResult(
+            K_PKE_CipherText cipherText, byte[] sharedSecret) {
+    }
+
+    /*
+    Key check functions from the beginning of sections 7.2 and 7.3 of the spec
+     */
+    public Object checkPublicKey(byte[] pk) throws InvalidKeyException {
+        //Encapsulation key type check
+        if (pk.length != mlKem_k * 384 + 32) {
+            throw new InvalidKeyException("Public key is not the correct size");
+        }
+
+        //Encapsulation key modulus check
+        int x, y, z, a, b;
+        for (int i = 0; i < mlKem_k * 384; i += 3) {
+            x = pk[i] & 0xFF;
+            y = pk[i + 1] & 0xFF;
+            z = pk[i + 2] & 0xFF;
+            a = x + ((y & 0xF) << 8);
+            b = (y >> 4) + (z << 4);
+            if ((a >= ML_KEM_Q) || (b >= ML_KEM_Q)) {
+                throw new InvalidKeyException("Coefficients in public key not in specified range");
+            }
+        }
+        return null;
```

Comment:
This is because the `NamedKEM` methods for checking keys can optionally return the key. I can pass that key along if that makes more sense

Reply:
Yes, pass the key along seems to make more sense.

Reply:
I'm going to leave this returning null for now because I believe the intention here is to have the function optionally return a parsed key. Since ML-KEM does not have parsed key objects this wouldn't make sense. With ML-DSA however it would.